README
README for Cache::Memcached::Managed

Copyright (c) 2005, 2006 BOOKINGS
Copyright (c) 2007, 2008 BOOKING.COM
Copyright (c) 2012 Elizabeth Mattijsen

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

Version:
 0.23

Required Modules:
 Cache::Memcached (any)

perl Makefile.PL
make
make test
make install


CHANGELOG
0.23    24 May 2012
        Fixed thinko reported by James M.  Added tests to test for thinko.
        Also fixed specifying a blessed object as data / directory source,
        although this may well break with non-hashref based implemenations.
        But it didn't work before anyway, so it at least now matches the
        documentation.

        Ensured it runs with 5.16.0, threaded and unthreaded.

0.22    22 April 2012
        Some more spelling fixes.  Added license information + adapted copyright
        message to reflect that I've taken it back in maintenance in my own
        time again.

0.21    31 March 2012
        Added spelling fixes as suggested by Nicholas Bamber, as suggested
        by Gregor Herrmann.  Finally released to CPAN, thanks to the impetus
        given by the QA Hackathon 2012 in Paris.

        24 May 2008
        Shielded some code from using Cache::Memcached specific internal
        methods, which do not appear to exist in Cache::Memcached::Fast.
        This at least let the test-suite complete, albeit with errors if
        the Cache::Memcached::Fast class is being used.  This probably will
        need some more work and some direct interaction with the developers
        of Cache::Memcached::Fast.  Testing with a class other than
        Cache::Memcached can be accomplished by setting the environment
        variable CACHE_MEMCACHE to the intended class, e.g.

          $ CACHE_MEMCACHED=Cache::Memcached::Fast make test

0.20    19 May 2008
        Added preliminary "memcached_class" parameter to "new", to allow other
        implementations of Cache::Memcached (such as Cache::Memcached::Fast)
        to be used.

0.19    8 October 2007
        Argh, 0.18 broke on 5.005 because of new style open().  Now rephrased        
        to old-style open() using globs.

0.18    7 October 2007
        Changed Makefile.PL to quit before attempting to install to kill all
        cpan-tester messages where Cache::Memcached::Managed is being tested
        without "memcached" being available and executable.  No changes made
        to the code.

0.17    3 October 2007
        Methods "add", "decr", "incr", "replace" and "set" now accept a third
        unnamed parameter for expiration.  Suggested by John Goulah.

        Fixed test-suite: most problems occurred because "localhost"
        apparently doesn't resolve on most systems anymore, so it couldn't
        find a port to work on.  Now using 127.0.0.1.

        Several other problems occurred because of additional statistics in
        newer versions of memcached.  Marked tests as TODO for now.

        Apparently the handling of dead servers has been improved, so the
        fetch on a dead servers wasn't failing, which was failing the test
        and subsequent statistics tests.  Also marked tests as TODO for now.

        Added "Incompatibility with Cache module" section to CAVEATS.

0.16    2 February 2007
        Updated the test-suite to support version 1.2.1 of memcached.  Brought
        code of test-suite up to PBP standards (code itself will follow in a
        later version).

        Load testing is no longer skipped on OS X, as the latest libevent
        / memcached combo seems to compile and install cleanly with all
        necessary workarounds installed automagically.  Instead a warning
        is displayed.

        Changed documentation of "flush_all" to point out that the
        "flush_interval.patch" is no longer necessary when using memcached
        1.2.1.

0.15    11 June 2006
        Cache::Memcached changed some of its internals with version 1.16.
        This version now supports both versions before and after.  Spotted
        by islue.

0.14    19 October 2005
        Update VERSION file (forgotten to do that in 0.13)
        Added fix for missing Kwalitee point for not using 'strict'

0.13    23 August 2005
        Added support for Cache::Memcached::Managed::Multi, to allow accessing
        multiple managed servers at the same time.

0.12    18 July 2005
        Added "inactive" parameter to "new".  Added
        Cache::Memcached::Managed::Inactive module for the "inactive"
        functionality.  This allows you to make your code in such a way that
        it doesn't make a difference whether or not there is an actual
        memcached server available.

0.11    8 July 2005
        Fixed problem in "group" and "get_multi" when using a non-standard
        delimiter.  Spotted by Torbjørn Køhle

0.10    6 July 2005
        Changed return value of "flush_all" to boolean rather than number
        of servers flushed.  Added support for "flush_interval".  Added
        file "flush_interval.patch" to distribution to add time flush_all
        support to memcached itself.

0.09    5 July 2005
        Further refined the fork checking / dead server logic: connection
        attempts are now made again after 20..30 seconds, instead of never
        in the current process.

        Fixed problem that would cause an empty list to be returned in case
        of failure in list context: this affected "set", "add", "replace",
        "incr" and "decr".  A false value is now returned regardless of
        context.

        4 July 2005
        Completely reworked the fork checking logic.  Child processes of
        parent processes that have been disconnected, now attempt to
        reconnect.

0.08    4 July 2005
        Fixed many documentation problems.  Added default data server
        functionality to "new".  Added "W" to allow expirations to be
        specified by week.

0.07    3 July 2005
        Fixed problem with return value of "start".  Optimized "stats".
        Added stress test related to groups, which unfortunately seems to
        fail on Tiger (with its crippled libevent support).

0.06    1 July 2005
        Fixed problems with relative keys and version support.

0.05    30 June 2005
        Added initial tests for checking behaviour after fork().


lib/Cache/Memcached/Managed/Inactive.pm
package Cache::Memcached::Managed::Inactive;

# Make sure we have version info for this module

$VERSION= '0.23';

#--------------------------------------------------------------------------
BEGIN { # We're fooling the Kwalitee checker into thinking we're strict
use strict;
}

#--------------------------------------------------------------------------
# No, we're NOT using strict here.  There are several reasons, the most
# important is that we're doing a lot of nasty stuff here.
# If you _do_ want stricture as a developer of this module, simply activate
# the line below here
#--------------------------------------------------------------------------
#use strict;

# Singleton object

my $self;

# At compile time
#  Create accessors returning undef

BEGIN {
    *$_ = sub { undef } foreach qw(
 add
 data
 decr
 delete
 delete_group
 delimiter
 directory
 expiration
 flush_all
 flush_interval
 get
 incr
 namespace
 replace
 reset
 set
 start
 stop
    );

#  Create accessors returning hash ref

    *$_ = sub { {} } foreach qw(
 errors
 get_group
 get_multi
 grab_group
 group
 stats
 version
    );

#  Create accessors returning list or hash ref

    *$_ = sub { wantarray ? () : {} } foreach qw(
 dead
 group_names
 servers
    );
} #BEGIN

# Satisfy -require-

1;

#---------------------------------------------------------------------------
#
# Class methods
#
#---------------------------------------------------------------------------
# new
#
# Return instantiated object
#
#  IN: 1 class
#      2..N hash with parameters
# OUT: 1 instantiated object

sub new { $self ||= bless {},shift } #new

#---------------------------------------------------------------------------
#
# Object methods
#
#---------------------------------------------------------------------------
# inactive
#
#  IN: 1 instantiated object
# OUT: 1 true

sub inactive { 1 } #inactive

#---------------------------------------------------------------------------

__END__

=head1 NAME

Cache::Memcached::Managed::Inactive - inactive Cache::Memcache::Managed object

=head1 SYNOPSIS

 use Cache::Memcached::Managed::Inactive;

 my $cache = Cache::Memcached::Managed::Inactive->new;

=head1 DESCRIPTION

Provides the same API as L<Cache::Memcached::Managed>, but doesn't do anything.

=head1 AUTHOR

 Elizabeth Mattijsen

=head1 COPYRIGHT

(C) 2005 - 2006 BOOKINGS
(C) 2007 BOOKING.COM

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

=cut


lib/Cache/Memcached/Managed/Multi.pm
package Cache::Memcached::Managed::Multi;

# Make sure we have version info for this module

$VERSION= '0.23';

BEGIN { # We're fooling the Kwalitee checker into thinking we're strict
use strict;
}

#--------------------------------------------------------------------------
# No, we're NOT using strict here.  There are several reasons, the most
# important is that we're doing a lot of nasty stuff here.
# If you _do_ want stricture as a developer of this module, simply activate
# the line below here
#--------------------------------------------------------------------------
#use strict;

# At compile time
#  Create list with method names

BEGIN {
    my @method = qw(
 add
 data
 dead
 decr
 delete
 delete_group
 delimiter
 directory
 errors
 expiration
 flush_all
 flush_interval
 get
 get_group
 get_multi
 grab_group
 group
 group_names
 inactive
 incr
 namespace
 replace
 reset
 servers
 set
 start
 stats
 stop
 version
    );

#  Create a method for each of the methods which
#   Obtains the list of objects
#   If called in list context
#    Call the methods in list context and return list of list refs
#   Elseif called in scalar context
#    Call the methods in scalar context and return list ref of values
#   Call the methods in void context

    eval <<SUB foreach @method;
sub $_ {
    my \$objects = shift;
    if (wantarray) {
        return map { [\$_->$_( \@_ )] } \@{\$objects};
    } elsif (defined wantarray) {
        return [map { scalar \$_->$_( \@_ ) } \@{\$objects}];
    }
    \$_->$_( \@_ ) foreach \@{\$objects};
} #$_
SUB
} #BEGIN

# Satisfy -require-

1;

#---------------------------------------------------------------------------
#
# Class methods
#
#---------------------------------------------------------------------------
# new
#
# Return instantiated object
#
#  IN: 1 class
#      2..N list of Cache::Memcached::Managed (compatible) objects
# OUT: 1 instantiated object

sub new { my $class = shift; bless \@_,$class } #new

#---------------------------------------------------------------------------
#
# Object methods
#
#---------------------------------------------------------------------------
# objects
#
#  IN: 1 instantiated object
# OUT: 1..N objects of which this object consists

sub objects { @{$_[0]} } #objects

#---------------------------------------------------------------------------

__END__

=head1 NAME

Cache::Memcached::Managed::Multi - multiple Cache::Memcache::Managed objects

=head1 SYNOPSIS

 use Cache::Memcached::Managed::Multi;

 my $multi = Cache::Memcached::Managed::Multi->new( @managed );

=head1 DESCRIPTION

Provides the same API as L<Cache::Memcached::Managed>, but applies all methods
called to all of the objects specified, except for L<new> and L<objects>.

=head1 CONTEXT

All methods are called on all of the L<Cache::Memcached::Managed> objects
in the same context (list, scalar or void) in which the method is called on
the L<Cache::Memcached::Managed::Multi> object.  The return value differs in
format depending on the context also:

=over 2

=item scalar

 my $listref = $multi->method;
 print "Result: @{$listref}\n";

When called in scalar context, a list ref with scalar values is returned in
the same order in which the objects are used (which is determined by the
order in which they were supplied with L<new> and returned by L<objects>..

=item list

 my @listref = $multi->method;
 print "Result $_: @{$listref[$_]}\n" foreach 0..$#listref;

When called in list context, a list of list references is returned in
the same order in which the objects are used (which is determined by the
order in which they were supplied with L<new> and returned by L<objects>.

=item void

 $multi->method;

When called in void context, nothing is returned (not strangely enough ;-).

=back

=head1 SPECIFIC CLASS METHODS

There is only one specific class method.

=head2 new

 my $multi = Cache::Memcached::Managed::Multi->new( @managed );

Create an object containing multiple L<Cache::Memcached::Managed> objects.
Returns the instantiated object.

=head1 SPECIFIC INSTANCE METHODS

=head2 objects

 my @managed = $multi->objects;

Returns the list of instantiated L<Cache::Memcached::Managed> objects that
the object is a proxy for.

=head1 AUTHOR

 Elizabeth Mattijsen

=head1 COPYRIGHT

(C) 2005 - 2006 BOOKINGS
(C) 2007 BOOKING.COM

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

=cut
